utils/aggregate_selector.sql

```diff
@@ -0,0 +1,208 @@
+-- Given the name of a root hypertable, the wanted downsampling interval, and a list of aggregate choices,
+-- this function will pick the largest-interval available compatible hypertable or continuous aggregate
+-- compatible with the downsampling interval and aggregate "with_column" constraints.
+--
+-- The returned JSONB will be the chosen aggregate object joined with extra metadata for the chosen
+-- table_schema, table_name and table_interval.
+CREATE OR REPLACE FUNCTION aggregate_choice(hypertable TEXT, selection_interval INTERVAL, 
+	aggregate_types JSONB, hypertable_schema TEXT DEFAULT 'public')
+RETURNS JSONB
+LANGUAGE SQL
+STABLE
+AS $BODY$
+	WITH available_tables AS (
+		SELECT table_name, table_schema, get_cagg_interval(table_name, table_schema) AS table_interval, get_jsonb_table_columns(table_name, table_schema) AS table_columns
+		FROM get_hypertable_caggs(hypertable, hypertable_schema)
+
+		UNION ALL
+
+		SELECT hypertable AS table_name, hypertable_schema AS table_schema, INTERVAL '0s' AS table_interval, get_jsonb_table_columns(hypertable, hypertable_schema) AS table_columns
+	), available_aggregates AS (
+		SELECT *
+		FROM jsonb_array_elements(aggregate_types) AS j (aggregate_option)
+	)
+	SELECT jsonb_build_object('table_name', table_name,
+		'table_schema', table_schema,
+		'table_interval', table_interval,
+		'interval_matched', table_interval = selection_interval) || aggregate_option
+	FROM available_tables, available_aggregates, (VALUES(1),(-1)) AS swp(interval_swap)
+	WHERE table_columns@>(aggregate_option->'with_columns') AND (table_interval*interval_swap) <= selection_interval
+	ORDER BY (table_interval*interval_swap) DESC, coalesce((aggregate_option->>'priority')::int) DESC
+	LIMIT 1
+$BODY$;
+
+
+-- This function will perform automatic downsampling of data
+-- from a given hypertable OR compatible continuous aggregate of the table.
+-- See top of the file for more extensive documentation.
+--
+-- The parameters are as follows:
+-- - hypertable: Name of the root hypertable to use.
+-- - data_interval: Interval that will be downsampled to using time_bucket
+-- - aggregate_choices: JSONB array of Objects of configurations for possible downsampling methods
+-- - groupby_clause: Clauses that will be inserted to the SELECT and GROUP BY query. Note that
+--   this may not be left empty. Instead, a dummy "true" or "0" can be inserted to ignore grouping.
+-- - filter_query: A clause inserted to the SELECT query after the FROM clause. Used to perform
+--   optional JOINs as well as the WHERE clause. MUST include a WHERE filter clause for the time column!
+-- - hypertable_schema: Schema of the hypertable, default 'public'
+-- - time_column: Name of the time column, default 'time'
+CREATE OR REPLACE FUNCTION ts_autoagg(hypertable TEXT, data_interval INTERVAL,
+	aggregate_choices JSONB, 
+	groupby_clause TEXT, filter_query TEXT,
+	hypertable_schema TEXT DEFAULT 'public',
+	time_column TEXT DEFAULT 'time')
```

Comment:
On thought, would it make sense to add maybe a debug flag and do a raise notice on the actual query that is run? 
That way the user could just grab it and run explain analyze and see what is going on. The function ends up hiding that.

Reply:
Yes, I've noticed that it's rather tricky to figure out what is happening inside a function...

I suppose I could reformat the query a bit to first generate a string and then optionally raise that string as debug flag if a value is set. It's not too much of a change and could definitely really help...

Thanks!

Reply:
If you have a moment, try the new pushed change.
I hope it works right